In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
df = pd.read_csv('smartphone_data.csv')

In [3]:
df

Product name    Price Specification Score  \
0                            Realme P4 Pro 5G  ₹24,999        73Spec Score   
1                                Realme P4 5G  ₹18,499        77Spec Score   
2                                 Vivo V60 5G  ₹36,999        86Spec Score   
3                          Xiaomi Redmi 15 5G  ₹14,999        78Spec Score   
4                  Motorola Moto G86 Power 5G  ₹18,999        83Spec Score   
...                                       ...      ...                 ...   
1012                            Realme 10 Pro  ₹17,990        80Spec Score   
1013            Realme 9 4G (8GB RAM + 128GB)  ₹15,990        82Spec Score   
1014                       Samsung Galaxy F13  ₹10,999        71Spec Score   
1015  Samsung Galaxy A33 5G (8GB RAM + 128GB)  ₹28,500        83Spec Score   
1016                               Poco M4 5G  ₹12,999        74Spec Score   

                                                    Sim  \
0        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
1        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1012                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
1013                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1014                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1015            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
1016     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   

                                              Processor  \
0               Snapdragon 7 Gen 4, Octa Core Processor   
1     Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   
2       Snapdragon 7 Gen4, Octa Core, 2.8 GHz Processor   
3      Snapdragon 6s Gen3, Octa Core, 2.3 GHz Processor   
4          Dimensity 7400, Octa Core, 2.6 GHz Processor   
...                                                 ...   
1012       Snapdragon 695, Octa Core, 2.2 GHz Processor   
1013       Snapdragon 680, Octa Core, 2.4 GHz Processor   
1014             Exynos 850, Octa Core, 2 GHz Processor   
1015         Exynos  1280, Octa Core, 2.4 GHz Processor   
1016     Dimensity 700 5G, Octa Core, 2.2 GHz Processor   

               RAM and Storage                                  Battery  \
0     8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1     6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
2     8 GB RAM, 128 GB inbuilt  6500 mAh Battery with 90W Fast Charging   
3     6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 33W Fast Charging   
4     8 GB RAM, 128 GB inbuilt  6720 mAh Battery with 33W Fast Charging   
...                        ...                                      ...   
1012  6 GB RAM, 128 GB inbuilt  5000 mAh Battery with 33W Fast Charging   
1013  8 GB RAM, 128 GB inbuilt  5000 mAh Battery with 33W Fast Charging   
1014   4 GB RAM, 64 GB inbuilt  6000 mAh Battery with 15W Fast Charging   
1015  8 GB RAM, 128 GB inbuilt  5000 mAh Battery with 25W Fast Charging   
1016   4 GB RAM, 64 GB inbuilt  5000 mAh Battery with 18W Fast Charging   

                                                Display  \
0     6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1     6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   
2     6.77 inches, 1080 x 2392 px, 120 Hz Display wi...   
3     6.9 inches, 1080 x 2340 px, 144 Hz Display wit...   
4     6.7 inches, 1220 x 2712 px, 120 Hz Display wit...   
...                                                 ...   
1012  6.72 inches, 1080 x 2400 px, 120 Hz Display wi...   
1013  6.4 inches, 1080 x 2400 px, 90 Hz Display with...   
1014  6.6 inches, 2408 x 1080 px Display with Water ...   
1015  6.4 inches, 1080 x 2400 px, 90 Hz Display with...   
1016  6.58 inches, 1080 x 2400 px, 90 Hz Display wit...   

                                                 Camera  \
0        

# Data Assessing
## Data quality issues are :- `completeness`, `validity`,`Accuracy`,`Consistency`

### Quality Issues

1. **model** - some brands are written diiferently like OPPO and LAVA in model column `consistency`
2. **price** - has unneccesary '₹' and "," between numbers `validity`
3. **Specification score** - has 'Spec score' after numbers `validity`
4. **Product name** - row '674' MTRJT is feature phone
5. **RAM** - row '539' and '275' has only ROM RAM is not given `completeness`
6. **Battery** - row '328','219','170','159','144','126','102','94','95','68','49','43', has no charger info `completeness`
7. **camera** row - 184,189,287,424,430,433,447,473,499,534,555,589,591,599,602,665,666,705,707,709,743,764,765,766,769,796,875,876,886,899,929,934,950,978,979  `validity`
8. **display** - sometimes frequency is not available `completeness`
9. row '496','660' = battery column has display info and disply column has camera info and camera column has memory card info and memory card column has OS info
11. **memory card** and **OS** - has scattered info
12. datatype of price and Specification score] is incorrect `validity`
13. There are 5 duplicate rows are present

### Tidiness Issues

1. **sim** - can be split into 3 cols has_5g, has_NFC, has_IR_Blaster
2. **ram** - can be split into 2 cols RAM and ROM
3. **processor** - can be split into processor name, cores and cpu speed.
4. **battery** - can be split into battery capacity, fast_charging_available
5. **display** - can be split into size, resolution_width, resolution_height and frequency
6. **camera** - can be split into front and rear camera
7. **card** - can be split into supported, extended_upto

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017 entries, 0 to 1016
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Product name         1017 non-null   object
 1   Price                1017 non-null   object
 2   Specification Score  1017 non-null   object
 3   Sim                  1017 non-null   object
 4   Processor            1017 non-null   object
 5   RAM and Storage      1017 non-null   object
 6   Battery              1017 non-null   object
 7   Display              1017 non-null   object
 8   Camera               1017 non-null   object
 9   Memory Card          1017 non-null   object
 10  OS                   1017 non-null   object
dtypes: object(11)
memory usage: 87.5+ KB


In [5]:
df.describe()

Product name    Price Specification Score  \
count                      1017     1017                1017   
unique                     1012      501                  51   
top     Apple iPhone 16 (512GB)  ₹14,999        74Spec Score   
freq                          2       20                  71   

                                       Sim  \
count                                 1017   
unique                                  14   
top     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
freq                                   315   

                                           Processor  \
count                                           1017   
unique                                           202   
top     Dimensity 6300, Octa Core, 2.4 GHz Processor   
freq                                             111   

                 RAM and Storage                                  Battery  \
count                       1017                                     1017   
unique                        28                                      169   
top     8 GB RAM, 128 GB inbuilt  5000 mAh Battery with 18W Fast Charging   
freq                         229                                       83   

                                                  Display  \
count                                                1017   
unique                                                207   
top     6.67 inches, 1080 x 2400 px, 120 Hz Display wi...   
freq                                                   75   

                                            Camera                Memory Card  \
count                                         1017                       1017   
unique                                         192                         36   
top     50 MP + 2 MP Dual Rear & 8 MP Front Camera  Memory Card Not Supported   
freq                                            77                        178   

                 OS  
count          1017  
unique           16  
top     Android v14  
freq            283

# Make a copy of dataset

In [6]:
df1 = df.copy()

# Handling "Price" column

In [7]:
df1['Price'] = df1['Price'].str.replace("₹","").str.replace(",","")

In [8]:
df1.head(1)

Product name  Price Specification Score  \
0  Realme P4 Pro 5G  24999        73Spec Score   

                                              Sim  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   

                                 Processor           RAM and Storage  \
0  Snapdragon 7 Gen 4, Octa Core Processor  8 GB RAM, 128 GB inbuilt   

                                   Battery  \
0  7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   

                                        Camera  Memory Card           OS  
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Android v15  No FM Radio

## Changing datatype of "Price" column 

In [9]:
df1['Price'] = df1['Price'].astype(int)

In [10]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017 entries, 0 to 1016
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Product name         1017 non-null   object
 1   Price                1017 non-null   int64 
 2   Specification Score  1017 non-null   object
 3   Sim                  1017 non-null   object
 4   Processor            1017 non-null   object
 5   RAM and Storage      1017 non-null   object
 6   Battery              1017 non-null   object
 7   Display              1017 non-null   object
 8   Camera               1017 non-null   object
 9   Memory Card          1017 non-null   object
 10  OS                   1017 non-null   object
dtypes: int64(1), object(10)
memory usage: 87.5+ KB


## Handling "Specification score" column

In [11]:
df1['Specification Score'] = df1['Specification Score'].str.replace('Spec Score',"").astype(int)

In [12]:
df1

Product name  Price  Specification Score  \
0                            Realme P4 Pro 5G  24999                   73   
1                                Realme P4 5G  18499                   77   
2                                 Vivo V60 5G  36999                   86   
3                          Xiaomi Redmi 15 5G  14999                   78   
4                  Motorola Moto G86 Power 5G  18999                   83   
...                                       ...    ...                  ...   
1012                            Realme 10 Pro  17990                   80   
1013            Realme 9 4G (8GB RAM + 128GB)  15990                   82   
1014                       Samsung Galaxy F13  10999                   71   
1015  Samsung Galaxy A33 5G (8GB RAM + 128GB)  28500                   83   
1016                               Poco M4 5G  12999                   74   

                                                    Sim  \
0        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
1        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
2     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
4                    Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
...                                                 ...   
1012                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
1013                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1014                     Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1015            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
1016     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   

                                              Processor  \
0               Snapdragon 7 Gen 4, Octa Core Processor   
1     Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   
2       Snapdragon 7 Gen4, Octa Core, 2.8 GHz Processor   
3      Snapdragon 6s Gen3, Octa Core, 2.3 GHz Processor   
4          Dimensity 7400, Octa Core, 2.6 GHz Processor   
...                                                 ...   
1012       Snapdragon 695, Octa Core, 2.2 GHz Processor   
1013       Snapdragon 680, Octa Core, 2.4 GHz Processor   
1014             Exynos 850, Octa Core, 2 GHz Processor   
1015         Exynos  1280, Octa Core, 2.4 GHz Processor   
1016     Dimensity 700 5G, Octa Core, 2.2 GHz Processor   

               RAM and Storage                                  Battery  \
0     8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1     6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
2     8 GB RAM, 128 GB inbuilt  6500 mAh Battery with 90W Fast Charging   
3     6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 33W Fast Charging   
4     8 GB RAM, 128 GB inbuilt  6720 mAh Battery with 33W Fast Charging   
...                        ...                                      ...   
1012  6 GB RAM, 128 GB inbuilt  5000 mAh Battery with 33W Fast Charging   
1013  8 GB RAM, 128 GB inbuilt  5000 mAh Battery with 33W Fast Charging   
1014   4 GB RAM, 64 GB inbuilt  6000 mAh Battery with 15W Fast Charging   
1015  8 GB RAM, 128 GB inbuilt  5000 mAh Battery with 25W Fast Charging   
1016   4 GB RAM, 64 GB inbuilt  5000 mAh Battery with 18W Fast Charging   

                                                Display  \
0     6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1     6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   
2     6.77 inches, 1080 x 2392 px, 120 Hz Display wi...   
3     6.9 inches, 1080 x 2340 px, 144 Hz Display wit...   
4     6.7 inches, 1220 x 2712 px, 120 Hz Display wit...   
...                                                 ...   
1012  6.72 inches, 1080 x 2400 px, 120 Hz Display wi...   
1013  6.4 inches, 1080 x 2400 px, 90 Hz Display with...   
1014  6.6 inches, 2408 x 1080 px Display with Water ...   
1015  6.4 inches, 1080 x 2400 px, 90 Hz Display with...   
1016  6.58 inches, 1080 x 2400 px, 90 Hz Display wit...   

                                                 Camera  \
0           50 MP + 8

In [13]:
df1 = df1.reset_index()

In [14]:
df1['index']=df1['index'] + 2

In [15]:
df1.head()

index                Product name  Price  Specification Score  \
0      2            Realme P4 Pro 5G  24999                   73   
1      3                Realme P4 5G  18499                   77   
2      4                 Vivo V60 5G  36999                   86   
3      5          Xiaomi Redmi 15 5G  14999                   78   
4      6  Motorola Moto G86 Power 5G  18999                   83   

                                                 Sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
2  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
4                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           Processor  \
0            Snapdragon 7 Gen 4, Octa Core Processor   
1  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   
2    Snapdragon 7 Gen4, Octa Core, 2.8 GHz Processor   
3   Snapdragon 6s Gen3, Octa Core, 2.3 GHz Processor   
4       Dimensity 7400, Octa Core, 2.6 GHz Processor   

            RAM and Storage                                  Battery  \
0  8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
2  8 GB RAM, 128 GB inbuilt  6500 mAh Battery with 90W Fast Charging   
3  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 33W Fast Charging   
4  8 GB RAM, 128 GB inbuilt  6720 mAh Battery with 33W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   
2  6.77 inches, 1080 x 2392 px, 120 Hz Display wi...   
3  6.9 inches, 1080 x 2340 px, 144 Hz Display wit...   
4  6.7 inches, 1220 x 2712 px, 120 Hz Display wit...   

                                              Camera  \
0        50 MP + 8 MP Dual Rear & 50 MP Front Camera   
1        50 MP + 8 MP Dual Rear & 16 MP Front Camera   
2  50 MP + 50 MP + 8 MP Triple Rear & 50 MP Front...   
3                50 MP Dual Rear & 8 MP Front Camera   
4        50 MP + 8 MP Dual Rear & 32 MP Front Camera   

                        Memory Card           OS  
0                       Android v15  No FM Radio  
1                       Android v15  No FM Radio  
2         Memory Card Not Supported  Android v15  
3  Memory Card Supported, upto 1 TB  Android v15  
4   Memory Card (Hybrid), upto 1 TB  Android v15

In [16]:
ram_rows = set((275,539))
battery_rows = set((496,660))
camera_rows = set((184,189,287,424,430,433,447,473,499,534,555,589,591,599,602,665,666,705,707,709,743,764,765,766,769,796,875,876,886,899,929,934,950,978,979 ))

In [17]:
df1[df1['index'].isin(ram_rows | battery_rows | camera_rows )]

index                                  Product name   Price  \
182    184                       Samsung Galaxy Z Fold 7  158999   
187    189                                Lava Agni 3 5G   15299   
273    275                          Apple iPhone 16 Plus   81900   
285    287      Samsung Galaxy Z Fold 7 (16GB RAM + 1TB)  216999   
422    424                        Motorola Razr 60 Ultra   89998   
428    430              Lava Agni 3 5G (8GB RAM + 256GB)   19999   
431    433                              Vivo X Fold 5 5G  144990   
445    447                    Samsung Galaxy Z Flip 7 5G  109999   
471    473                       Samsung Galaxy Z Fold 6  115720   
494    496                               Apple iPhone 12   44999   
497    499    Samsung Galaxy Z Fold 7 (12GB RAM + 512GB)  186999   
532    534                              Motorola Razr 60   50229   
537    539                  Apple iPhone 16 Plus (256GB)   92900   
553    555                       Google Pixel 9 Pro Fold  134999   
587    589                             Vivo X Fold 3 Pro  159999   
589    591                        Tecno Phantom V Fold 2   89999   
597    599                    Samsung Galaxy Z Flip 6 5G   80999   
600    602                              Motorola Razr 50   48898   
658    660                       Apple iPhone 12 (128GB)   43999   
663    665  Samsung Galaxy Z Flip 7 FE (8GB RAM + 256GB)   78399   
664    666   Samsung Galaxy Z Flip 7 5G (12GB RAM+512GB)  100999   
703    705              Lava Blaze Duo (8GB RAM + 128GB)   17999   
705    707                                Lava Blaze Duo   16999   
707    709                 Samsung Galaxy Z Flip 7 FE 5G   89999   
741    743                  OnePlus Open 5G Apex Edition  149998   
762    764      Samsung Galaxy Z Fold 6 (12GB RAM + 1TB)  200999   
763    765    Samsung Galaxy Z Fold 6 (12GB RAM + 512GB)  134999   
764    766               Samsung Galaxy Z Flip 6 (512GB)   79499   
767    769                        Tecno Phantom V Flip 2   54999   
794    796                          Infinix Zero Flip 5G   44999   
873    875      Samsung Galaxy Z Fold 5 (12GB RAM + 1TB)  133000   
874    876    Samsung Galaxy Z Fold 5 (12GB RAM + 512GB)  107999   
884    886                        Motorola Razr 50 Ultra   59999   
897    899                             OPPO Find N3 Flip   69994   
927    929                          Tecno Phantom V Fold   70999   
932    934                       Samsung Galaxy Z Fold 5  109999   
948    950                       Samsung Galaxy Z Flip 5   69999   
976    978    Samsung Galaxy Z Fold 4 (12GB RAM + 512GB)   94990   
977    979    Samsung Galaxy Z Fold 4 (12GB RAM + 512GB)   94990   

     Specification Score                                                Sim  \
182                   95      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
187                   81           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
273                   82            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
285                   97      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
422                   93            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
428                   82           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
431                   94  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
445                   91      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
471                   90      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
494                   74            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
497                   95      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
532                   89            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
537                   83            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
553                   91            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
587                   96  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
589                 

In [18]:
df1[df1['index'].isin(ram_rows & battery_rows & camera_rows )]

Empty DataFrame
Columns: [index, Product name, Price, Specification Score, Sim, Processor, RAM and Storage, Battery, Display, Camera, Memory Card, OS]
Index: []

# Handling "Battery" column

In [19]:
df1[df1['index'].isin(battery_rows)]

index             Product name  Price  Specification Score  \
494    496          Apple iPhone 12  44999                   74   
658    660  Apple iPhone 12 (128GB)  43999                   74   

                                         Sim  \
494  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
658  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   

                                    Processor           RAM and Storage  \
494  Bionic A14, Hexa Core, 3.1 GHz Processor   4 GB RAM, 64 GB inbuilt   
658  Bionic A14, Hexa Core, 3.1 GHz Processor  4 GB RAM, 128 GB inbuilt   

                                               Battery  \
494  6.1 inches, 1170 x 2532 px Display with Large ...   
658  6.1 inches, 1170 x 2532 px Display with Large ...   

                                          Display                     Camera  \
494  12 MP + 12 MP Dual Rear & 12 MP Front Camera  Memory Card Not Supported   
658  12 MP + 12 MP Dual Rear & 12 MP Front Camera  Memory Card Not Supported   

    Memory Card           OS  
494     iOS v14  No FM Radio  
658     iOS v14  No FM Radio

In [20]:
temp_df = df1[df1['index'].isin(battery_rows)]

In [21]:
temp_df.iloc[:,7:].shift(1,axis=1)

Battery                                            Display  \
494    None  6.1 inches, 1170 x 2532 px Display with Large ...   
658    None  6.1 inches, 1170 x 2532 px Display with Large ...   

                                           Camera                Memory Card  \
494  12 MP + 12 MP Dual Rear & 12 MP Front Camera  Memory Card Not Supported   
658  12 MP + 12 MP Dual Rear & 12 MP Front Camera  Memory Card Not Supported   

          OS  
494  iOS v14  
658  iOS v14

In [22]:
x = temp_df.iloc[:,7:].shift(1,axis=1).values
print(x)

[[None
  '6.1 inches, 1170\u2009x\u20092532\u2009px Display with Large Notch'
  '12\u2009MP + 12\u2009MP Dual Rear & 12\u2009MP Front Camera'
  'Memory Card Not Supported' 'iOS v14']
 [None
  '6.1 inches, 1170\u2009x\u20092532\u2009px Display with Large Notch'
  '12\u2009MP + 12\u2009MP Dual Rear & 12\u2009MP Front Camera'
  'Memory Card Not Supported' 'iOS v14']]


In [23]:
df1.loc[temp_df.index,temp_df.columns[7:]] = x

## Rechecking

In [24]:
df1[df1['index'].isin(battery_rows)]

index             Product name  Price  Specification Score  \
494    496          Apple iPhone 12  44999                   74   
658    660  Apple iPhone 12 (128GB)  43999                   74   

                                         Sim  \
494  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
658  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   

                                    Processor           RAM and Storage  \
494  Bionic A14, Hexa Core, 3.1 GHz Processor   4 GB RAM, 64 GB inbuilt   
658  Bionic A14, Hexa Core, 3.1 GHz Processor  4 GB RAM, 128 GB inbuilt   

    Battery                                            Display  \
494    None  6.1 inches, 1170 x 2532 px Display with Large ...   
658    None  6.1 inches, 1170 x 2532 px Display with Large ...   

                                           Camera                Memory Card  \
494  12 MP + 12 MP Dual Rear & 12 MP Front Camera  Memory Card Not Supported   
658  12 MP + 12 MP Dual Rear & 12 MP Front Camera  Memory Card Not Supported   

          OS  
494  iOS v14  
658  iOS v14

# Handling "Camera" column

In [25]:
df1[df1['index'].isin(camera_rows)]

index                                  Product name   Price  \
182    184                       Samsung Galaxy Z Fold 7  158999   
187    189                                Lava Agni 3 5G   15299   
285    287      Samsung Galaxy Z Fold 7 (16GB RAM + 1TB)  216999   
422    424                        Motorola Razr 60 Ultra   89998   
428    430              Lava Agni 3 5G (8GB RAM + 256GB)   19999   
431    433                              Vivo X Fold 5 5G  144990   
445    447                    Samsung Galaxy Z Flip 7 5G  109999   
471    473                       Samsung Galaxy Z Fold 6  115720   
497    499    Samsung Galaxy Z Fold 7 (12GB RAM + 512GB)  186999   
532    534                              Motorola Razr 60   50229   
553    555                       Google Pixel 9 Pro Fold  134999   
587    589                             Vivo X Fold 3 Pro  159999   
589    591                        Tecno Phantom V Fold 2   89999   
597    599                    Samsung Galaxy Z Flip 6 5G   80999   
600    602                              Motorola Razr 50   48898   
663    665  Samsung Galaxy Z Flip 7 FE (8GB RAM + 256GB)   78399   
664    666   Samsung Galaxy Z Flip 7 5G (12GB RAM+512GB)  100999   
703    705              Lava Blaze Duo (8GB RAM + 128GB)   17999   
705    707                                Lava Blaze Duo   16999   
707    709                 Samsung Galaxy Z Flip 7 FE 5G   89999   
741    743                  OnePlus Open 5G Apex Edition  149998   
762    764      Samsung Galaxy Z Fold 6 (12GB RAM + 1TB)  200999   
763    765    Samsung Galaxy Z Fold 6 (12GB RAM + 512GB)  134999   
764    766               Samsung Galaxy Z Flip 6 (512GB)   79499   
767    769                        Tecno Phantom V Flip 2   54999   
794    796                          Infinix Zero Flip 5G   44999   
873    875      Samsung Galaxy Z Fold 5 (12GB RAM + 1TB)  133000   
874    876    Samsung Galaxy Z Fold 5 (12GB RAM + 512GB)  107999   
884    886                        Motorola Razr 50 Ultra   59999   
897    899                             OPPO Find N3 Flip   69994   
927    929                          Tecno Phantom V Fold   70999   
932    934                       Samsung Galaxy Z Fold 5  109999   
948    950                       Samsung Galaxy Z Flip 5   69999   
976    978    Samsung Galaxy Z Fold 4 (12GB RAM + 512GB)   94990   
977    979    Samsung Galaxy Z Fold 4 (12GB RAM + 512GB)   94990   

     Specification Score                                                Sim  \
182                   95      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
187                   81           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
285                   97      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
422                   93            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
428                   82           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
431                   94  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
445                   91      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
471                   90      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
497                   95      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
532                   89            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
553                   91            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
587                   96  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
589                   95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
597                   90      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
600                   88            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
663                   90      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
664                   91      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
703                   79           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
705                   77           Dual Sim, 3G, 4G, 5G

In [26]:
temp_df = df1[df1['index'].isin(camera_rows)]

In [27]:
temp_df = temp_df[~temp_df['Camera'].str.contains('MP')]

In [28]:
df1.loc[temp_df.index,'Camera'] = temp_df['Memory Card'].values

## Rechecking

In [29]:
df1[df1['index'].isin(camera_rows)]

index                                  Product name   Price  \
182    184                       Samsung Galaxy Z Fold 7  158999   
187    189                                Lava Agni 3 5G   15299   
285    287      Samsung Galaxy Z Fold 7 (16GB RAM + 1TB)  216999   
422    424                        Motorola Razr 60 Ultra   89998   
428    430              Lava Agni 3 5G (8GB RAM + 256GB)   19999   
431    433                              Vivo X Fold 5 5G  144990   
445    447                    Samsung Galaxy Z Flip 7 5G  109999   
471    473                       Samsung Galaxy Z Fold 6  115720   
497    499    Samsung Galaxy Z Fold 7 (12GB RAM + 512GB)  186999   
532    534                              Motorola Razr 60   50229   
553    555                       Google Pixel 9 Pro Fold  134999   
587    589                             Vivo X Fold 3 Pro  159999   
589    591                        Tecno Phantom V Fold 2   89999   
597    599                    Samsung Galaxy Z Flip 6 5G   80999   
600    602                              Motorola Razr 50   48898   
663    665  Samsung Galaxy Z Flip 7 FE (8GB RAM + 256GB)   78399   
664    666   Samsung Galaxy Z Flip 7 5G (12GB RAM+512GB)  100999   
703    705              Lava Blaze Duo (8GB RAM + 128GB)   17999   
705    707                                Lava Blaze Duo   16999   
707    709                 Samsung Galaxy Z Flip 7 FE 5G   89999   
741    743                  OnePlus Open 5G Apex Edition  149998   
762    764      Samsung Galaxy Z Fold 6 (12GB RAM + 1TB)  200999   
763    765    Samsung Galaxy Z Fold 6 (12GB RAM + 512GB)  134999   
764    766               Samsung Galaxy Z Flip 6 (512GB)   79499   
767    769                        Tecno Phantom V Flip 2   54999   
794    796                          Infinix Zero Flip 5G   44999   
873    875      Samsung Galaxy Z Fold 5 (12GB RAM + 1TB)  133000   
874    876    Samsung Galaxy Z Fold 5 (12GB RAM + 512GB)  107999   
884    886                        Motorola Razr 50 Ultra   59999   
897    899                             OPPO Find N3 Flip   69994   
927    929                          Tecno Phantom V Fold   70999   
932    934                       Samsung Galaxy Z Fold 5  109999   
948    950                       Samsung Galaxy Z Flip 5   69999   
976    978    Samsung Galaxy Z Fold 4 (12GB RAM + 512GB)   94990   
977    979    Samsung Galaxy Z Fold 4 (12GB RAM + 512GB)   94990   

     Specification Score                                                Sim  \
182                   95      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
187                   81           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
285                   97      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
422                   93            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
428                   82           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
431                   94  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
445                   91      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
471                   90      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
497                   95      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
532                   89            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
553                   91            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
587                   96  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
589                   95            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
597                   90      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
600                   88            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
663                   90      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
664                   91      Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
703                   79           Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi   
705                   77           Dual Sim, 3G, 4G, 5G

# Handling "Memory Card" column

In [30]:
df1['Memory Card'].value_counts()

Memory Card
Memory Card Not Supported                                                  180
Memory Card Supported, upto 1 TB                                           157
Memory Card (Hybrid), upto 1 TB                                            153
Android v15                                                                103
Android v14                                                                 87
Memory Card (Hybrid), upto 2 TB                                             63
Memory Card (Hybrid)                                                        43
Memory Card Supported, upto 2 TB                                            42
Memory Card Supported                                                       38
Android v13                                                                 29
Android v12                                                                 29
Memory Card Supported, upto 512 GB                                          22
Memory Card Supported, upto 256 GB      

In [31]:
temp_df = df1[df1['Memory Card'].str.contains('MP')]

In [32]:
df1.loc[temp_df.index,'Memory Card'] = "Memory Card Not Supported"

In [33]:
df1['Memory Card'].value_counts()

Memory Card
Memory Card Not Supported             215
Memory Card Supported, upto 1 TB      157
Memory Card (Hybrid), upto 1 TB       153
Android v15                           103
Android v14                            87
Memory Card (Hybrid), upto 2 TB        63
Memory Card (Hybrid)                   43
Memory Card Supported, upto 2 TB       42
Memory Card Supported                  38
Android v13                            29
Android v12                            29
Memory Card Supported, upto 512 GB     22
Memory Card Supported, upto 256 GB     16
Memory Card (Hybrid), upto 1.5 TB      12
Android v16                             3
Memory Card (Hybrid), upto 256 GB       2
Android v10                             1
Memory Card (Hybrid), upto 128 GB       1
Android v10.1                           1
Name: count, dtype: int64

# Handling "OS" column

In [34]:
df1[~df1['Memory Card'].str.contains('Memory Card')]

index                           Product name  Price  \
0         2                       Realme P4 Pro 5G  24999   
1         3                           Realme P4 5G  18499   
11       13                        OPPO Reno 14 5G  37998   
12       14                           iQOO Neo 10R  24105   
13       15                      OnePlus Nord 5 5G  31999   
...     ...                                    ...    ...   
973     975        Vivo V25 Pro (12GB RAM + 256GB)  39999   
987     989                   Xiaomi Redmi K50i 5G  22999   
993     995     OnePlus Nord 2T (12GB RAM + 256GB)  30999   
1004   1006  Motorola Edge 30 5G (8GB RAM + 128GB)  23990   
1008   1010    Xiaomi 12 Pro 5G (12GB RAM + 256GB)  36499   

      Specification Score                                                Sim  \
0                      73     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
1                      77     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
11                     86     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
12                     81     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
13                     84  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
...                   ...                                                ...   
973                    83                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   
987                    80     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
993                    83            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
1004                   83            Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
1008                   87  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   

                                              Processor  \
0               Snapdragon 7 Gen 4, Octa Core Processor   
1     Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   
11        Dimensity 8350, Octa Core, 3.35 GHz Processor   
12       Snapdragon 8s Gen3, Octa Core, 3 GHz Processor   
13       Snapdragon 8s Gen3, Octa Core, 3 GHz Processor   
...                                                 ...   
973          Dimensity 1300, Octa Core, 3 GHz Processor   
987      Dimensity  8100, Octa Core, 2.85 GHz Processor   
993          Dimensity 1300, Octa Core, 3 GHz Processor   
1004     Snapdragon 778G+, Octa Core, 2.5 GHz Processor   
1008      Snapdragon 8 Gen1, Octa Core, 3 GHz Processor   

                RAM and Storage                                   Battery  \
0      8 GB RAM, 128 GB inbuilt   7000 mAh Battery with 80W Fast Charging   
1      6 GB RAM, 128 GB inbuilt   7000 mAh Battery with 80W Fast Charging   
11     8 GB RAM, 256 GB inbuilt   6000 mAh Battery with 80W Fast Charging   
12     8 GB RAM, 128 GB inbuilt   6400 mAh Battery with 80W Fast Charging   
13     8 GB RAM, 256 GB inbuilt   6800 mAh Battery with 80W Fast Charging   
...                         ...                                       ...   
973   12 GB RAM, 256 GB inbuilt   4830 mAh Battery with 66W Fast Charging   
987    6 GB RAM, 128 GB inbuilt   5080 mAh Battery with 67W Fast Charging   
993   12 GB RAM, 256 GB inbuilt   4500 mAh Battery with 80W Fast Charging   
1004   8 GB RAM, 128 GB inbuilt   4020 mAh Battery with 33W Fast Charging   
1008  12 GB RAM, 256 GB inbuilt  4600 mAh Battery with 120W Fast Charging   

                                                Display  \
0     6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1     6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   
11    6.59 inches, 1256 x 2760 px, 120 Hz Display wi...   
12    6.78 inches, 1260 x 2800 px, 144 Hz Display wi...   
13    6.83 inches, 1272 x 2800 px, 144 Hz Display wi...   
...                                                 ...   
973   6.56 inches, 1080 x 2376 px, 120 Hz Display wi...   
987   6.6 inches, 1080 x 2460 px, 144 Hz Display wit...   
993   6.43 inches, 1080 x 2400 px, 90 Hz Display wit...   
1004  6.5 inches, 1080 x 2400 px, 144 Hz Display wit...   
1008  6.73 inches, 1440 x 320

In [35]:
pd.set_option("display.max_rows",None)

In [36]:
temp_df = df1[~df1['Memory Card'].str.contains('Memory Card')]

In [37]:
df1.loc[temp_df.index,"OS"] = temp_df['Memory Card'].values

In [38]:
df1[~df1['Memory Card'].str.contains('Memory Card')]

index                                    Product name   Price  \
0         2                                Realme P4 Pro 5G   24999   
1         3                                    Realme P4 5G   18499   
11       13                                 OPPO Reno 14 5G   37998   
12       14                                    iQOO Neo 10R   24105   
13       15                               OnePlus Nord 5 5G   31999   
19       21                                        Oppo K13   17620   
24       26                            Motorola Edge 60 Pro   30169   
25       27                             OPPO Reno 14 Pro 5G   49998   
26       28                                Realme 15 Pro 5G   30800   
27       29                            Oppo F27 Pro Plus 5G   17999   
31       33           Realme GT Neo 3 5G (12GB RAM + 256GB)   17000   
38       40                                    Vivo Y400 5G   21999   
39       41                                    Realme P3 5G   16392   
40       42                                     iQOO Neo 10   31998   
45       47                                Nothing Phone 3a   23258   
48       50                                Vivo Y400 Pro 5G   23999   
51       53                                    Realme 15 5G   24999   
55       57                             Infinix Note 50s 5G   15999   
58       60                                       iQOO Z10x   13498   
59       61                  Realme P4 5G (8GB RAM + 256GB)   21499   
60       62              Realme P4 Pro 5G (8GB RAM + 256GB)   26999   
61       63          OPPO Reno 13 Pro 5G (12GB RAM + 512GB)   35198   
65       67                           Samsung Galaxy S24 5G   50400   
67       69                           Samsung Galaxy S25 5G   69750   
74       76             Realme P4 Pro 5G (12GB RAM + 256GB)   28999   
75       77                                Realme P3 Pro 5G   23999   
76       78           Oppo F27 Pro Plus 5G (8GB RAM +128GB)   15999   
79       81                           Samsung Galaxy A56 5G   38999   
82       84                                      iQOO 13 5G   54998   
84       86                  Realme P4 5G (8GB RAM + 128GB)   19499   
86       88                           Samsung Galaxy A36 5G   30999   
89       91                       Motorola Moto Edge 50 Neo   19999   
90       92                              Nothing Phone 3 5G   49728   
95       97         Motorola Edge 60 Pro (16GB RAM + 512GB)   37999   
99      101                                 OPPO Reno 13 5G   30990   
103     105                                  Poco X7 Pro 5G   23899   
104     106                                   iQOO Z10 Lite    9998   
105     107                  iQOO Neo 10R (8GB RAM + 256GB)   25948   
108     110                     Xiaomi Redmi Note 14 Pro 5G   20990   
110     112                      Oppo K13 (8GB RAM + 256GB)   19870   
113     115                             Google Pixel 10 Pro  109999   
117     119                         Samsung Galaxy S24 Plus   52999   
119     121                                   Realme X7 Pro   15000   
132     134                                 Oppo F29 Pro 5G   25999   
135     137                                      Poco X7 5G   17390   
139     141                                       Honor X9c   21999   
141     143                                  Xiaomi 14 Civi   39999   
143     145                                  Google Pixel 9   59880   
144     146                     Xiaomi Redmi Note 13 Pro 5G   19699   
145     147               OnePlus Nord 5 (12GB RAM + 256GB)   34999   
149     151                                   Vivo T3 Ultra   25599   
155     157                                       Xiaomi 15   64998   
162     164                  Realme P3 5G (8GB RAM + 256GB)   18499   
163     165                 iQOO Neo 10R (12GB RAM + 256GB)   27832   
165     167                                       Honor 200   22999   
166     168           Infinix 

In [39]:
df1.loc[temp_df.index,'Memory Card'] = "Memory Card Not Supported"

In [40]:
df1['Memory Card'].value_counts()

Memory Card
Memory Card Not Supported             468
Memory Card Supported, upto 1 TB      157
Memory Card (Hybrid), upto 1 TB       153
Memory Card (Hybrid), upto 2 TB        63
Memory Card (Hybrid)                   43
Memory Card Supported, upto 2 TB       42
Memory Card Supported                  38
Memory Card Supported, upto 512 GB     22
Memory Card Supported, upto 256 GB     16
Memory Card (Hybrid), upto 1.5 TB      12
Memory Card (Hybrid), upto 256 GB       2
Memory Card (Hybrid), upto 128 GB       1
Name: count, dtype: int64

In [41]:
df1['OS'].value_counts()

OS
Android v14                  370
Android v15                  278
Android v13                  164
Android v12                  113
Memory Card Not Supported     21
Android v11                   20
iOS v18                       17
Android v10                   10
iOS v17                        6
Android v16                    6
iOS v16                        5
iOS v14                        2
Android v10.1                  2
iOS v15                        1
iOS v13                        1
Bluetooth                      1
Name: count, dtype: int64

In [42]:
df1[df1['OS'] == 'Bluetooth']

index Product name  Price  Specification Score  \
672    674       MTR JT   1950                   25   

                                Sim        Processor  \
672  Dual Sim, 3G, 4G, VoLTE, Wi-Fi  1 MHz Processor   

              RAM and Storage           Battery  \
672  32 MB RAM, 32 MB inbuilt  3000 mAh Battery   

                              Display            Camera           Memory Card  \
672  2.8 inches, 480 x 640 px Display  Main Rear Camera  Memory Card (Hybrid)   

            OS  
672  Bluetooth

In [43]:
temp_df = df1[df1['OS'] == 'Bluetooth']

In [44]:
df1.loc[temp_df.index,'OS'] = np.nan

In [45]:
df1.head()

index                Product name  Price  Specification Score  \
0      2            Realme P4 Pro 5G  24999                   73   
1      3                Realme P4 5G  18499                   77   
2      4                 Vivo V60 5G  36999                   86   
3      5          Xiaomi Redmi 15 5G  14999                   78   
4      6  Motorola Moto G86 Power 5G  18999                   83   

                                                 Sim  \
0     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
1     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
2  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Bl...   
3     Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
4                 Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                           Processor  \
0            Snapdragon 7 Gen 4, Octa Core Processor   
1  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   
2    Snapdragon 7 Gen4, Octa Core, 2.8 GHz Processor   
3   Snapdragon 6s Gen3, Octa Core, 2.3 GHz Processor   
4       Dimensity 7400, Octa Core, 2.6 GHz Processor   

            RAM and Storage                                  Battery  \
0  8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
2  8 GB RAM, 128 GB inbuilt  6500 mAh Battery with 90W Fast Charging   
3  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 33W Fast Charging   
4  8 GB RAM, 128 GB inbuilt  6720 mAh Battery with 33W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   
2  6.77 inches, 1080 x 2392 px, 120 Hz Display wi...   
3  6.9 inches, 1080 x 2340 px, 144 Hz Display wit...   
4  6.7 inches, 1220 x 2712 px, 120 Hz Display wit...   

                                              Camera  \
0        50 MP + 8 MP Dual Rear & 50 MP Front Camera   
1        50 MP + 8 MP Dual Rear & 16 MP Front Camera   
2  50 MP + 50 MP + 8 MP Triple Rear & 50 MP Front...   
3                50 MP Dual Rear & 8 MP Front Camera   
4        50 MP + 8 MP Dual Rear & 32 MP Front Camera   

                        Memory Card           OS  
0         Memory Card Not Supported  Android v15  
1         Memory Card Not Supported  Android v15  
2         Memory Card Not Supported  Android v15  
3  Memory Card Supported, upto 1 TB  Android v15  
4   Memory Card (Hybrid), upto 1 TB  Android v15

In [46]:
temp_df = df1[df1['OS'] == 'Memory Card Not Supported']

In [47]:
df1.loc[temp_df.index,'OS'] = np.nan

In [48]:
df1['OS'].value_counts()

OS
Android v14      370
Android v15      278
Android v13      164
Android v12      113
Android v11       20
iOS v18           17
Android v10       10
iOS v17            6
Android v16        6
iOS v16            5
Android v10.1      2
iOS v14            2
iOS v15            1
iOS v13            1
Name: count, dtype: int64

In [49]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017 entries, 0 to 1016
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   index                1017 non-null   int64 
 1   Product name         1017 non-null   object
 2   Price                1017 non-null   int64 
 3   Specification Score  1017 non-null   int64 
 4   Sim                  1017 non-null   object
 5   Processor            1017 non-null   object
 6   RAM and Storage      1017 non-null   object
 7   Battery              1015 non-null   object
 8   Display              1017 non-null   object
 9   Camera               1017 non-null   object
 10  Memory Card          1017 non-null   object
 11  OS                   995 non-null    object
dtypes: int64(3), object(9)
memory usage: 95.5+ KB


In [50]:
df1.isnull().sum()

index                   0
Product name            0
Price                   0
Specification Score     0
Sim                     0
Processor               0
RAM and Storage         0
Battery                 2
Display                 0
Camera                  0
Memory Card             0
OS                     22
dtype: int64

In [51]:
df1['Processor'].value_counts()

Processor
Dimensity 6300, Octa Core, 2.4 GHz Processor                             111
Dimensity 6100 Plus, Octa Core, 2.2 GHz Processor                         27
Snapdragon 4 Gen2, Octa Core, 2.2 GHz Processor                           24
Dimensity 7300, Octa Core, 2.5 GHz Processor                              24
Snapdragon 7 Gen3, Octa Core, 2.63 GHz Processor                          21
Helio G85, Octa Core, 2 GHz Processor                                     19
Snapdragon 8s Gen3, Octa Core, 3 GHz Processor                            18
Snapdragon 7s Gen3, Octa Core, 2.5 GHz Processor                          18
Snapdragon 695, Octa Core, 2.2 GHz Processor                              17
Exynos 1380, Octa Core, 2.4 GHz Processor                                 15
Unisoc T612, Octa Core, 1.8 GHz Processor                                 15
Snapdragon 7s Gen2, Octa Core, 2.4 GHz Processor                          15
Dimensity 6020, Octa Core, 2.2 GHz Processor                      

# Removing feature phone data

In [52]:
df1[df1['Product name'] == 'MTR JT']

index Product name  Price  Specification Score  \
672    674       MTR JT   1950                   25   

                                Sim        Processor  \
672  Dual Sim, 3G, 4G, VoLTE, Wi-Fi  1 MHz Processor   

              RAM and Storage           Battery  \
672  32 MB RAM, 32 MB inbuilt  3000 mAh Battery   

                              Display            Camera           Memory Card  \
672  2.8 inches, 480 x 640 px Display  Main Rear Camera  Memory Card (Hybrid)   

      OS  
672  NaN

In [53]:
df1.drop(672,inplace=True)

## Rechecking

In [54]:
df1[df1['Product name'] == 'MTR JT']

Empty DataFrame
Columns: [index, Product name, Price, Specification Score, Sim, Processor, RAM and Storage, Battery, Display, Camera, Memory Card, OS]
Index: []

In [55]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1016 entries, 0 to 1016
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   index                1016 non-null   int64 
 1   Product name         1016 non-null   object
 2   Price                1016 non-null   int64 
 3   Specification Score  1016 non-null   int64 
 4   Sim                  1016 non-null   object
 5   Processor            1016 non-null   object
 6   RAM and Storage      1016 non-null   object
 7   Battery              1014 non-null   object
 8   Display              1016 non-null   object
 9   Camera               1016 non-null   object
 10  Memory Card          1016 non-null   object
 11  OS                   995 non-null    object
dtypes: int64(3), object(9)
memory usage: 103.2+ KB


# Create new calumn 'Brand name' from existing 'Product name' column

In [56]:
brand_name = df1['Product name'].str.split().str.get(0)

In [57]:
df1.insert(1, 'Brand name',brand_name)

In [58]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     Realme  Realme P4 Pro 5G  24999                   73   
1      3     Realme      Realme P4 5G  18499                   77   

                                              Sim  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   

                                           Processor  \
0            Snapdragon 7 Gen 4, Octa Core Processor   
1  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   

            RAM and Storage                                  Battery  \
0  8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15

# Convert all brand names into lower case

In [59]:
df1['Brand name'] = df1['Brand name'].str.lower()

In [60]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     realme  Realme P4 Pro 5G  24999                   73   
1      3     realme      Realme P4 5G  18499                   77   

                                              Sim  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster   

                                           Processor  \
0            Snapdragon 7 Gen 4, Octa Core Processor   
1  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   

            RAM and Storage                                  Battery  \
0  8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15

## Creating three new columns 'Has_5G', 'Has_NFC', 'Has_IR_Blaster' from existing column 'Sim'

In [61]:
has_5g = df1['Sim'].str.contains("5G")
has_nfc = df1['Sim'].str.contains("NFC")
has_ir_blaster = df1['Sim'].str.contains("IR Blaster")

In [62]:
df1.insert(6, "Has_5G", has_5g)
df1.insert(7, "Has_NFC", has_nfc)
df1.insert(8, "Has_IR_Blaster", has_ir_blaster)

In [63]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     realme  Realme P4 Pro 5G  24999                   73   
1      3     realme      Realme P4 5G  18499                   77   

                                              Sim  Has_5G  Has_NFC  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   

   Has_IR_Blaster                                          Processor  \
0            True            Snapdragon 7 Gen 4, Octa Core Processor   
1            True  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   

            RAM and Storage                                  Battery  \
0  8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15

## Creating three new columns 'Processor_name' , 'num_cores', 'processor_speed' from existing column 'Processor'

In [64]:
processor_name = df1['Processor'].str.split(",").str.get(0)
num_cores = df1['Processor'].str.split(",").str.get(1)
processor_speed = df1['Processor'].str.split(",").str.get(2)

In [65]:
df1.insert(10,'Processor_name',processor_name)
df1.insert(11,'num_cores',num_cores)
df1.insert(12,'Processor_speed',processor_speed)

In [66]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     realme  Realme P4 Pro 5G  24999                   73   
1      3     realme      Realme P4 5G  18499                   77   

                                              Sim  Has_5G  Has_NFC  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   

   Has_IR_Blaster                                          Processor  \
0            True            Snapdragon 7 Gen 4, Octa Core Processor   
1            True  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   

         Processor_name             num_cores     Processor_speed  \
0    Snapdragon 7 Gen 4   Octa Core Processor                 NaN   
1  Dimensity 7400 Ultra             Octa Core   2.6 GHz Processor   

            RAM and Storage                                  Battery  \
0  8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15

In [67]:
df1['Processor_name'].value_counts()

Processor_name
Dimensity 6300                            111
Dimensity 6100 Plus                        27
Snapdragon 4 Gen2                          24
Dimensity 7300                             24
Snapdragon 7 Gen3                          21
Helio G85                                  19
Snapdragon 7s Gen3                         18
Snapdragon 8s Gen3                         18
Snapdragon 695                             17
Unisoc T612                                16
Dimensity 7400                             15
Snapdragon 7s Gen2                         15
Dimensity 6020                             15
Exynos 1380                                15
Dimensity 7050                             14
Snapdragon 8 Gen3                          14
Snapdragon 6 Gen1                          14
Dimensity 6080                             14
Helio G36                                  12
Dimensity 7300 Ultimate                    11
Snapdragon 680                             11
Snapdragon 8 Elite 

In [68]:
df1['Processor_name'] = df1['Processor_name'].str.strip()

## shift the columns 'Processor_name' , 'num_cores', 'processor_speed' becouse it contains scattered info

In [69]:
df1[df1['Processor_name'].str.contains("Core")]

index Brand name                   Product name  Price  \
690    692      bigme            Bigme Hibreak E INK  42990   
694    696       itel  itel Zeno 10 (4GB RAM + 64GB)   5899   
698    700       itel                   itel Zeno 10   5999   
702    704       lava       Lava O3 (3GB RAM + 64GB)   5799   
800    802       lava                        Lava O3   5899   

     Specification Score                             Sim  Has_5G  Has_NFC  \
690                   64       Single Sim, 3G, 4G, Wi-Fi   False    False   
694                   56  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   False    False   
698                   55  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   False    False   
702                   55  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   False    False   
800                   57  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   False    False   

     Has_IR_Blaster                     Processor       Processor_name  \
690           False  Octa Core, 2.3 GHz Processor            Octa Core   
694           False           Octa Core Processor  Octa Core Processor   
698           False           Octa Core Processor  Octa Core Processor   
702           False           Octa Core Processor  Octa Core Processor   
800           False           Octa Core Processor  Octa Core Processor   

              num_cores Processor_speed           RAM and Storage  \
690   2.3 GHz Processor             NaN  6 GB RAM, 128 GB inbuilt   
694                 NaN             NaN   4 GB RAM, 64 GB inbuilt   
698                 NaN             NaN   3 GB RAM, 64 GB inbuilt   
702                 NaN             NaN   3 GB RAM, 64 GB inbuilt   
800                 NaN             NaN   4 GB RAM, 64 GB inbuilt   

                                     Battery  \
690                         3300 mAh Battery   
694      5000 mAh Battery with Fast Charging   
698      5000 mAh Battery with Fast Charging   
702  5000 mAh Battery with 10W Fast Charging   
800  5000 mAh Battery with 10W Fast Charging   

                                               Display  \
690                 5.84 inches, 720 x 1440 px Display   
694  6.6 inches, 720 x 1612 px Display with Water D...   
698  6.6 inches, 720 x 1612 px Display with Water D...   
702  6.75 inches, 1600 x 720 px Display with Water ...   
800  6.75 inches, 1600 x 720 px Display with Water ...   

                                  Camera                       Memory Card  \
690       13 MP Rear & 5 MP Front Camera  Memory Card Supported, upto 1 TB   
694   8 MP Dual Rear & 5 MP Front Camera             Memory Card Supported   
698   8 MP Dual Rear & 5 MP Front Camera             Memory Card Supported   
702  13 MP Dual Rear & 5 MP Front Camera   Memory Card (Hybrid), upto 1 TB   
800  13 MP Dual Rear & 5 MP Front Camera   Memory Card (Hybrid), upto 1 TB   

              OS  
690  Android v11  
694  Android v14  
698  Android v14  
702  Android v14  
800  Android v14

In [70]:
temp_df = df1[df1['Processor_name'].str.contains("Core")][["Processor_name", "num_cores", "Processor_speed"]].shift(1, axis=1)

In [71]:
temp_df

Processor_name            num_cores     Processor_speed
690           None            Octa Core   2.3 GHz Processor
694           None  Octa Core Processor                 NaN
698           None  Octa Core Processor                 NaN
702           None  Octa Core Processor                 NaN
800           None  Octa Core Processor                 NaN

In [72]:
df1.loc[temp_df.index,["Processor_name", "num_cores", "Processor_speed"]] = temp_df.values

In [73]:
df1[df1['Processor_name'] == "9863A"] 

index Brand name          Product name  Price  Specification Score  \
667    669  blackzone  Blackzone Warrior 4G   6799                   49   

                                Sim  Has_5G  Has_NFC  Has_IR_Blaster  \
667  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   False    False           False   

                      Processor Processor_name             num_cores  \
667  9863A, Octa Core Processor          9863A   Octa Core Processor   

    Processor_speed           RAM and Storage           Battery  \
667             NaN  4 GB RAM, 128 GB inbuilt  5250 mAh Battery   

                                               Display  \
667  6.6 inches, 1560 x 720 px Display with Large N...   

                             Camera                Memory Card           OS  
667  16 MP Rear & 8 MP Front Camera  Memory Card Not Supported  Android v13

In [74]:
df1.loc[667, 'Processor_name'] = "Unisoc SC9863A"

In [75]:
df1.loc[667]

index                                                                669
Brand name                                                     blackzone
Product name                                        Blackzone Warrior 4G
Price                                                               6799
Specification Score                                                   49
Sim                                       Dual Sim, 3G, 4G, VoLTE, Wi-Fi
Has_5G                                                             False
Has_NFC                                                            False
Has_IR_Blaster                                                     False
Processor                                     9863A, Octa Core Processor
Processor_name                                            Unisoc SC9863A
num_cores                                            Octa Core Processor
Processor_speed                                                      NaN
RAM and Storage                                 4 G

## Creating new column "Processor_brand" from existing column "Processor_name"

In [76]:
processor_brand = df1['Processor_name'].str.split(' ').str.get(0).str.lower()

In [77]:
df1.insert(11,"Processor_brand",processor_brand)

In [78]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     realme  Realme P4 Pro 5G  24999                   73   
1      3     realme      Realme P4 5G  18499                   77   

                                              Sim  Has_5G  Has_NFC  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   

   Has_IR_Blaster                                          Processor  \
0            True            Snapdragon 7 Gen 4, Octa Core Processor   
1            True  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   

         Processor_name Processor_brand             num_cores  \
0    Snapdragon 7 Gen 4      snapdragon   Octa Core Processor   
1  Dimensity 7400 Ultra       dimensity             Octa Core   

      Processor_speed           RAM and Storage  \
0                 NaN  8 GB RAM, 128 GB inbuilt   
1   2.6 GHz Processor  6 GB RAM, 128 GB inbuilt   

                                   Battery  \
0  7000 mAh Battery with 80W Fast Charging   
1  7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15

In [79]:
df1['num_cores'].value_counts()

num_cores
 Octa Core              930
 Hexa Core               32
 Octa Core Processor     25
 Deca Core                9
 Quad Core                6
Octa Core Processor       4
 Nine Cores               2
 Nine-Cores               2
 Nine Core                2
Octa Core                 1
Name: count, dtype: int64

In [80]:
df1['num_cores'] = df1['num_cores'].str.strip()

In [81]:
df1['num_cores'].value_counts()

num_cores
Octa Core              931
Hexa Core               32
Octa Core Processor     29
Deca Core                9
Quad Core                6
Nine-Cores               2
Nine Cores               2
Nine Core                2
Name: count, dtype: int64

# Replace values "Octa Core Processor"  with "Octa Core" and "Nine-Cores", "Nine Cores" with "Nine Core"

In [82]:
df1['num_cores'] = df1['num_cores'].str.replace("Octa Core Processor","Octa Core").str.replace("Nine-Cores","Nine Core").str.replace("Nine Cores","Nine Core")

In [83]:
df1['num_cores'].value_counts()

num_cores
Octa Core    960
Hexa Core     32
Deca Core      9
Nine Core      6
Quad Core      6
Name: count, dtype: int64

In [84]:
df1['Processor_speed'].value_counts()

Processor_speed
2.4 GHz Processor     207
2.2 GHz Processor     137
2.5 GHz Processor     106
2 GHz Processor        64
2.6 GHz Processor      45
2.3 GHz Processor      43
3 GHz Processor        39
3.2 GHz Processor      39
1.8 GHz Processor      36
2.8 GHz Processor      29
3.35 GHz Processor     25
2.63 GHz Processor     21
3.25 GHz Processor     17
3.3 GHz Processor      16
3.1 GHz Processor      16
1.6 GHz Processor      15
4.47 GHz Processor     13
4.32 GHz Processor     11
4.04 GHz Processor     10
1.82 GHz Processor      9
2.85 GHz Processor      9
2.7 GHz Processor       8
3.05 GHz Processor      8
3.36 GHz Processor      7
4.05 GHz Processor      7
3.63 GHz Processor      6
3.22 GHz Processor      6
3.46 GHz Processor      5
2.84 GHz Processor      4
2.9 GHz Processor       4
3.4 GHz Processor       4
3.39 GHz Processor      3
3.11 GHz Processor      2
2.1 GHz Processor       2
512 MHz Processor       2
2.96 GHz Processor      2
2.05 GHz Processor      2
3.78 GHz Processor    

In [85]:
df1['Processor_speed'].str.strip().str.split().str.get(0)

0        NaN
1        2.6
2        2.8
3        2.3
4        2.6
5        2.3
6       3.35
7        2.5
8        2.6
9       3.25
10       2.5
11      3.35
12         3
13         3
14       2.5
15       2.5
16       2.3
17      3.11
18       2.4
19       2.3
20       2.6
21       2.5
22       2.3
23       3.3
24      3.35
25      3.25
26       2.8
27       2.6
28       3.2
29       2.5
30      2.63
31      2.85
32       2.8
33      3.35
34      4.47
35       2.4
36       2.6
37       2.8
38       2.2
39       2.3
40       3.2
41       2.6
42       2.4
43       3.3
44       2.6
45       2.5
46       2.4
47       2.3
48       2.5
49       2.5
50      3.25
51       NaN
52      2.63
53       2.4
54       2.7
55       2.5
56       2.3
57       2.3
58       2.5
59       2.6
60       NaN
61      3.35
62       3.2
63       2.3
64       2.5
65       3.2
66      4.04
67      4.47
68      4.32
69       2.3
70       2.6
71      3.25
72       2.5
73       2.4
74       NaN
75       2.5
76       2.6

In [86]:
df1['Processor_speed'] = df1['Processor_speed'].str.strip().str.split().str.get(0)

## Changing the datatype of "Processor_speed" column

In [87]:
df1['Processor_speed'].dtype

dtype('O')

In [88]:
df1['Processor_speed'] = df1['Processor_speed'].astype(float)

In [89]:
df1['Processor_speed'].dtype

dtype('float64')

## Rename "Processor_speed" column

In [90]:
df1.rename(columns={"Processor_speed" : "Processor_speed(GHz)"},inplace=True)

In [91]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     realme  Realme P4 Pro 5G  24999                   73   
1      3     realme      Realme P4 5G  18499                   77   

                                              Sim  Has_5G  Has_NFC  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   

   Has_IR_Blaster                                          Processor  \
0            True            Snapdragon 7 Gen 4, Octa Core Processor   
1            True  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...   

         Processor_name Processor_brand  num_cores  Processor_speed(GHz)  \
0    Snapdragon 7 Gen 4      snapdragon  Octa Core                   NaN   
1  Dimensity 7400 Ultra       dimensity  Octa Core                   2.6   

            RAM and Storage                                  Battery  \
0  8 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   
1  6 GB RAM, 128 GB inbuilt  7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15

## Creating new column "Ram_capacity" from existing column 'RAM and Storage'

In [92]:
ram_capacity = df1['RAM and Storage'].str.strip().str.split(",").str.get(0).str.replace("RAM","")

In [93]:
df1.insert(15,"Ram_capacity",ram_capacity)

In [94]:
internal_storage = df1['RAM and Storage'].str.strip().str.split(",").str.get(1).str.replace("inbuilt","")

In [95]:
df1.insert(16,"Internal_storage",internal_storage)

In [96]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     realme  Realme P4 Pro 5G  24999                   73   
1      3     realme      Realme P4 5G  18499                   77   

                                              Sim  Has_5G  Has_NFC  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   

   Has_IR_Blaster                                          Processor  ...  \
0            True            Snapdragon 7 Gen 4, Octa Core Processor  ...   
1            True  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...  ...   

   num_cores Processor_speed(GHz)           RAM and Storage  Ram_capacity  \
0  Octa Core                  NaN  8 GB RAM, 128 GB inbuilt         8 GB    
1  Octa Core                  2.6  6 GB RAM, 128 GB inbuilt         6 GB    

  Internal_storage                                  Battery  \
0          128 GB   7000 mAh Battery with 80W Fast Charging   
1          128 GB   7000 mAh Battery with 80W Fast Charging   

                                             Display  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15  

[2 rows x 22 columns]

## Clean the data in "Ram_capacity" column by replacing ('128 GB inbuilt', '256 GB inbuilt' and '512 GB inbuilt') these values by np.nan

In [97]:
df1['Ram_capacity'].value_counts()

Ram_capacity
8 GB              422
4 GB              189
12 GB             183
6 GB              147
16 GB              34
3 GB               26
2 GB               11
24 GB               1
128 GB inbuilt      1
256 GB inbuilt      1
512 GB inbuilt      1
Name: count, dtype: int64

In [98]:
df1[df1['Ram_capacity'] == "128 GB inbuilt"]

index Brand name          Product name  Price  Specification Score  \
273    275      apple  Apple iPhone 16 Plus  81900                   82   

                                         Sim  Has_5G  Has_NFC  Has_IR_Blaster  \
273  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC    True     True           False   

                                     Processor  ...  num_cores  \
273  Bionic A18, Hexa Core, 4.04 GHz Processor  ...  Hexa Core   

    Processor_speed(GHz) RAM and Storage    Ram_capacity Internal_storage  \
273                 4.04  128 GB inbuilt  128 GB inbuilt              NaN   

                                 Battery  \
273  4674 mAh Battery with Fast Charging   

                                               Display  \
273  6.7 inches, 1290 x 2796 px Display with Dynami...   

                                           Camera                Memory Card  \
273  48 MP + 12 MP Dual Rear & 12 MP Front Camera  Memory Card Not Supported   

          OS  
273  iOS v18  

[1 rows x 22 columns]

In [99]:
df1.loc[273,'Internal_storage'] = "128 GB"

In [100]:
df1.loc[537,'Internal_storage'] = "256 GB"

In [101]:
df1.loc[714,'Internal_storage'] = "512 GB"

In [102]:
df1.loc[273,'Ram_capacity'] = np.nan
df1.loc[537,'Ram_capacity'] = np.nan
df1.loc[714,'Ram_capacity'] = np.nan

In [103]:
df1.loc[714]

index                                                                 716
Brand name                                                          apple
Product name                                 Apple iPhone 16 Plus (512GB)
Price                                                              109999
Specification Score                                                    84
Sim                               Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC
Has_5G                                                               True
Has_NFC                                                              True
Has_IR_Blaster                                                      False
Processor                       Bionic A18, Hexa Core, 4.04 GHz Processor
Processor_name                                                 Bionic A18
Processor_brand                                                    bionic
num_cores                                                       Hexa Core
Processor_speed(GHz)                  

In [104]:
df1['Internal_storage'] = df1['Internal_storage'].str.strip()

In [105]:
df1['Internal_storage'].value_counts()

Internal_storage
128 GB    468
256 GB    319
64 GB     112
512 GB     87
32 GB      18
1 TB        9
16 GB       3
Name: count, dtype: int64

## Create new columns "Battery_capacity" and "Fast_charging" from "Battery" column

In [106]:
battery_capacity = df1['Battery'].str.strip().str.split("with").str.get(0).str.split().str.get(0).astype(float)

In [107]:
df1.insert(18,"Battery_capacity",battery_capacity)

In [108]:
fast_charging = df1['Battery'].str.strip().str.split("with").str.get(1).str.split().str.get(0).str.replace("W","")

In [109]:
df1.insert(19,"Fast_charging(W)",fast_charging)

## Fill the null values with "-1" (means fast charging is not available)

In [110]:
df1['Fast_charging(W)'].isnull().sum()

np.int64(27)

In [111]:
df1['Fast_charging(W)'] = df1['Fast_charging(W)'].fillna('-1')

In [112]:
df1['Fast_charging(W)'].isnull().sum()

np.int64(0)

## Replace "Fast" with "0" (means Fast charging is available but "W" (charging speed) is not mentioned)

In [113]:
df1['Fast_charging(W)'] = df1['Fast_charging(W)'].str.replace("Fast","0").astype(int)

In [114]:
df1['Fast_charging(W)'].value_counts()

Fast_charging(W)
 45     142
 80     115
 25     104
 18     103
 33      94
 15      66
 10      52
 44      50
 0       46
 90      44
 67      36
 120     35
-1       27
 100     24
 30      16
 68      15
 66      11
 70       8
 50       7
 65       6
 20       4
 37       2
 27       2
 35       2
 125      2
 150      1
 23       1
 5        1
Name: count, dtype: int64

## Creating new columns "Screen_size(inches)" , "Resolution(px)", "Refresh_rate(Hz)" from "Display" column

In [115]:
screen_size = df1['Display'].str.strip().str.split(",").str.get(0).str.split().str.get(0).astype(float)

In [116]:
df1.insert(21,"Screen_size(inches)",screen_size)

In [117]:
resolution = df1['Display'].str.strip().str.split(",").str.get(1).str.strip().str.split("px").str.get(0)

In [118]:
df1.insert(22,"Resolution(px)",resolution)

In [119]:
refresh_rate = df1['Display'].str.strip().str.split(",").str.get(2).str.strip().str.split("Hz").str.get(0).astype(float)

In [120]:
df1.insert(23,"Refresh_rate(Hz)",refresh_rate)

## We replace null values with "60" because we know the standard refresh rate is 60

In [121]:
df1['Refresh_rate(Hz)'].isnull().sum()

np.int64(121)

In [122]:
df1['Refresh_rate(Hz)'] = df1['Refresh_rate(Hz)'].fillna(60)

In [123]:
df1['Refresh_rate(Hz)'].isnull().sum()

np.int64(0)

In [124]:
df1.head(2)

index Brand name      Product name  Price  Specification Score  \
0      2     realme  Realme P4 Pro 5G  24999                   73   
1      3     realme      Realme P4 5G  18499                   77   

                                              Sim  Has_5G  Has_NFC  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster    True    False   

   Has_IR_Blaster                                          Processor  ...  \
0            True            Snapdragon 7 Gen 4, Octa Core Processor  ...   
1            True  Dimensity 7400 Ultra, Octa Core, 2.6 GHz Proce...  ...   

                                   Battery Battery_capacity Fast_charging(W)  \
0  7000 mAh Battery with 80W Fast Charging           7000.0               80   
1  7000 mAh Battery with 80W Fast Charging           7000.0               80   

                                             Display Screen_size(inches)  \
0  6.8 inches, 1280 x 2800 px, 144 Hz Display wit...                6.80   
1  6.77 inches, 1080 x 2392 px, 144 Hz Display wi...                6.77   

  Resolution(px) Refresh_rate(Hz)  \
0   1280 x 2800             144.0   
1   1080 x 2392             144.0   

                                        Camera                Memory Card  \
0  50 MP + 8 MP Dual Rear & 50 MP Front Camera  Memory Card Not Supported   
1  50 MP + 8 MP Dual Rear & 16 MP Front Camera  Memory Card Not Supported   

            OS  
0  Android v15  
1  Android v15  

[2 rows x 27 columns]

## Creating new columns from "Camera" column

In [125]:
def camera_extractor(text):
    if "Quad" in text:
        return 4
    elif "Triple" in text:
        return 3
    elif "Dual" in text:
        return 2
    elif 'Missing' in text:
        return 'Missing'
    else:
        return '1'

In [126]:
no_rear_cameras = df1['Camera'].str.strip().str.split("&").str.get(0).apply(camera_extractor)

In [127]:
df1.insert(25, "Num_rear_cameras",no_rear_cameras)

In [128]:
no_front_cameras = df1['Camera'].str.strip().str.split("&").str.get(1).fillna("Missing").apply(camera_extractor)

In [129]:
df1.insert(26, "Num_front_cameras",no_front_cameras)

In [130]:
pd.set_option('display.max_columns',None)

In [131]:
primary_rear_camera = df1['Camera'].str.strip().str.split("&").str.get(0).str.split("MP").str.get(0)

In [132]:
df1.insert(27,"Primary_rear_camera(MP)", primary_rear_camera)

In [133]:
primary_front_camera = df1['Camera'].str.strip().str.split("&").str.get(1).str.split("MP").str.get(0)

In [134]:
df1.insert(28,"Primary_front_camera(MP)", primary_front_camera)

## Replacing "Missing" with np.nan

In [135]:
df1["Num_front_cameras"] = df1["Num_front_cameras"].replace("Missing",np.nan)

In [136]:
df1["Primary_front_camera(MP)"] = df1["Primary_front_camera(MP)"].str.strip()

## Replacing " Main Front Camera" with values by searching

In [144]:
df1[df1["Primary_front_camera(MP)"] == "Main Front Camera"]

Empty DataFrame
Columns: [index, Brand name, Product name, Price, Specification Score, Sim, Has_5G, Has_NFC, Has_IR_Blaster, Processor, Processor_name, Processor_brand, num_cores, Processor_speed(GHz), RAM and Storage, Ram_capacity, Internal_storage, Battery, Battery_capacity, Fast_charging(W), Display, Screen_size(inches), Resolution(px), Refresh_rate(Hz), Camera, Num_rear_cameras, Num_front_cameras, Primary_rear_camera(MP), Primary_front_camera(MP), Memory Card, OS]
Index: []

In [145]:
df1.loc[[549,697],"Primary_front_camera(MP)"] = [8,2]

## Rechecking

In [146]:
df1[df1["Primary_front_camera(MP)"] == " Main Front Camera"]

Empty DataFrame
Columns: [index, Brand name, Product name, Price, Specification Score, Sim, Has_5G, Has_NFC, Has_IR_Blaster, Processor, Processor_name, Processor_brand, num_cores, Processor_speed(GHz), RAM and Storage, Ram_capacity, Internal_storage, Battery, Battery_capacity, Fast_charging(W), Display, Screen_size(inches), Resolution(px), Refresh_rate(Hz), Camera, Num_rear_cameras, Num_front_cameras, Primary_rear_camera(MP), Primary_front_camera(MP), Memory Card, OS]
Index: []

In [151]:
df1["Primary_front_camera(MP)"] = df1["Primary_front_camera(MP)"].astype(float)

In [153]:
df1["Memory Card"].value_counts()

Memory Card
Memory Card Not Supported             468
Memory Card Supported, upto 1 TB      157
Memory Card (Hybrid), upto 1 TB       153
Memory Card (Hybrid), upto 2 TB        63
Memory Card Supported, upto 2 TB       42
Memory Card (Hybrid)                   42
Memory Card Supported                  38
Memory Card Supported, upto 512 GB     22
Memory Card Supported, upto 256 GB     16
Memory Card (Hybrid), upto 1.5 TB      12
Memory Card (Hybrid), upto 256 GB       2
Memory Card (Hybrid), upto 128 GB       1
Name: count, dtype: int64

## Creating new column "Extended_memory" from "Memory Card" column

In [158]:
df1['Extended_memory'] = df1['Memory Card'].str.strip().str.split("upto").str.get(-1)

In [161]:
df1['Extended_memory'] = df1['Extended_memory'].str.replace("Memory Card (Hybrid)","Not Specified").str.replace("Memory Card Supported","Not Specified")

In [172]:
df1['Extended_memory'] = df1['Extended_memory'].str.replace("Memory Card Not Supported","0").str.strip()

In [173]:
df1['Extended_memory'].value_counts()

Extended_memory
0                468
1 TB             310
2 TB             105
Not Specified     80
512 GB            22
256 GB            18
1.5 TB            12
128 GB             1
Name: count, dtype: int64

## Replacing values in "OS" column

In [163]:
df1['OS'].value_counts()

OS
Android v14      370
Android v15      278
Android v13      164
Android v12      113
Android v11       20
iOS v18           17
Android v10       10
iOS v17            6
Android v16        6
iOS v16            5
Android v10.1      2
iOS v14            2
iOS v15            1
iOS v13            1
Name: count, dtype: int64

In [166]:
df1["OS"] = df1['OS'].str.strip().str.split().str.get(0)

In [167]:
df1['OS'].value_counts()

OS
Android    963
iOS         32
Name: count, dtype: int64

In [168]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1016 entries, 0 to 1016
Data columns (total 32 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     1016 non-null   int64  
 1   Brand name                1016 non-null   object 
 2   Product name              1016 non-null   object 
 3   Price                     1016 non-null   int64  
 4   Specification Score       1016 non-null   int64  
 5   Sim                       1016 non-null   object 
 6   Has_5G                    1016 non-null   bool   
 7   Has_NFC                   1016 non-null   bool   
 8   Has_IR_Blaster            1016 non-null   bool   
 9   Processor                 1016 non-null   object 
 10  Processor_name            1011 non-null   object 
 11  Processor_brand           1011 non-null   object 
 12  num_cores                 1013 non-null   object 
 13  Processor_speed(GHz)      984 non-null    float64
 14  RAM and Stora

In [174]:
df1.columns

Index(['index', 'Brand name', 'Product name', 'Price', 'Specification Score',
       'Sim', 'Has_5G', 'Has_NFC', 'Has_IR_Blaster', 'Processor',
       'Processor_name', 'Processor_brand', 'num_cores',
       'Processor_speed(GHz)', 'RAM and Storage', 'Ram_capacity',
       'Internal_storage', 'Battery', 'Battery_capacity', 'Fast_charging(W)',
       'Display', 'Screen_size(inches)', 'Resolution(px)', 'Refresh_rate(Hz)',
       'Camera', 'Num_rear_cameras', 'Num_front_cameras',
       'Primary_rear_camera(MP)', 'Primary_front_camera(MP)', 'Memory Card',
       'OS', 'Extended_memory'],
      dtype='object')

## Drop the unnecessary columns

In [177]:
export_df = df1.drop(columns=['index','Sim','Processor','RAM and Storage','Battery','Display','Camera','Memory Card'])

In [178]:
export_df.head(2)

Brand name      Product name  Price  Specification Score  Has_5G  Has_NFC  \
0     realme  Realme P4 Pro 5G  24999                   73    True    False   
1     realme      Realme P4 5G  18499                   77    True    False   

   Has_IR_Blaster        Processor_name Processor_brand  num_cores  \
0            True    Snapdragon 7 Gen 4      snapdragon  Octa Core   
1            True  Dimensity 7400 Ultra       dimensity  Octa Core   

   Processor_speed(GHz) Ram_capacity Internal_storage  Battery_capacity  \
0                   NaN        8 GB            128 GB            7000.0   
1                   2.6        6 GB            128 GB            7000.0   

   Fast_charging(W)  Screen_size(inches) Resolution(px)  Refresh_rate(Hz)  \
0                80                 6.80   1280 x 2800              144.0   
1                80                 6.77   1080 x 2392              144.0   

  Num_rear_cameras Num_front_cameras Primary_rear_camera(MP)  \
0                2                 1                     50    
1                2                 1                     50    

   Primary_front_camera(MP)       OS Extended_memory  
0                      50.0  Android               0  
1                      16.0  Android               0

## Rename the columns

In [181]:
export_df.rename(columns={"Product name":"Model" , "Brand name":"Brand_name", "Specification Score":"Rating","Battery_capacity":"Battery_capacity(mAh)"},inplace=True)

In [184]:
export_df.rename(columns={"Processor_name":"Processor"},inplace=True)

In [185]:
export_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1016 entries, 0 to 1016
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Brand_name                1016 non-null   object 
 1   Model                     1016 non-null   object 
 2   Price                     1016 non-null   int64  
 3   Rating                    1016 non-null   int64  
 4   Has_5G                    1016 non-null   bool   
 5   Has_NFC                   1016 non-null   bool   
 6   Has_IR_Blaster            1016 non-null   bool   
 7   Processor                 1011 non-null   object 
 8   Processor_brand           1011 non-null   object 
 9   num_cores                 1013 non-null   object 
 10  Processor_speed(GHz)      984 non-null    float64
 11  Ram_capacity              1013 non-null   object 
 12  Internal_storage          1016 non-null   object 
 13  Battery_capacity(mAh)     1014 non-null   float64
 14  Fast_charging

In [186]:
export_df.to_csv('Smartphone_cleaned_v2.csv',index=False)